### ISSUE TITLE
gossipsub: `multi_hop_propagation` test failed in CI

### ISSUE THREAD

**Issue by thomaseizinger** [libp2p/rust-libp2p#3216]
## Summary

See https://github.com/libp2p/rust-libp2p/actions/runs/3652352144/jobs/6170648691#step:10:673.

```
test multi_hop_propagation ... FAILED

failures:

error: test failed, to rerun pass `-p libp2p-gossipsub --test smoke`
---- multi_hop_propagation stdout ----
thread 'multi_hop_propagation' panicked at '[quickcheck] TEST FAILED (runtime error). Arguments: (44, 3353857924130305067)
Error: Timed out waiting for all nodes to receive the msg but only have 0/44.', /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/quickcheck-1.0.3/src/tester.rs:165:28
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

## Actual behaviour

<details><summary>Debug Output</summary>
<p>

```
<output>
```
</p>
</details>

## Version

- libp2p version (version number, commit, or branch):

## Would you like to work on fixing this bug?

No.

**Comment by thomaseizinger** [comment]
Another instance: https://github.com/libp2p/rust-libp2p/actions/runs/3966238163/jobs/6796926824

**Comment by tcoratger** [comment]
@thomaseizinger On the last CI I had, it seems to work:

```
running 1 test
test multi_hop_propagation ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

Is there any problem with this test?

**Comment by thomaseizinger** [comment]
It is failing sometimes unfortunately.

I haven't had time yet to investigate.

**Comment by tcoratger** [comment]
@thomaseizinger Sure, tell me if there is something to bring here :)

**Comment by thomaseizinger** [comment]
> @thomaseizinger Sure, tell me if there is something to bring here :)

I'd be great to figure out why it fails. Feel free to dig into it :)
When I had a brief look last time, it looked difficult to figure out unfortunately but I could be wrong.

**Comment by mxinden** [comment]
Another instance: https://github.com/libp2p/rust-libp2p/actions/runs/5206276713/jobs/9392670573

**Comment by thomaseizinger** [comment]
Another instance: https://github.com/libp2p/rust-libp2p/actions/runs/5376198296/jobs/9753168660?pr=4114

**Comment by mxinden** [comment]
Another instance: https://github.com/libp2p/rust-libp2p/actions/runs/6324923977/job/17175397377?pr=4535

**Comment by bochaco** [comment]
Hi folks, I'm running into an issue of having a network where a subset of nodes don't seem to become aware of the topics the rest of nodes are subscribed to and yet not forwarding the msg through out others. So wondering if this may be the same reason/root-cause of the `multi_hop_propagation` test failing as the name seems to imply...?...

More specifically, if I have 10 peers, with only 1 peer subscribed to a topic, usually only ~6 of them would be aware of it and therefore be able to publish the msg successfully on that topic. However if published from any of the other 3 peers, they won't route the msg neither directly (since they aren't aware of any peer subscribed to it) nor through any of the other ~6 peers (who are aware of the topic). 
I thought that by having the `flood_publish` flag set msgs. would be always forwarded through other peers?
I'm just looking for any thoughts/advise/suggestions from any of you.

**Comment by thomaseizinger** [comment]
Thanks for the report!

I've not looked into the details of this test but assuming it has a decent assertion, I'd conclude that it seems to work _sometimes_. Are you saying you are experiencing the same thing in production that subscriptions only sometimes get propagated?

If that is the case, then it could also almost certainly be a bug and not just a problem with the test (which I had assumed until now).

cc @AgeManning Have you ever encountered this?

**Comment by AgeManning** [comment]
Yeah. So the `multi_hop_propagation` test is non-deterministic. We can change some of the parameters in there to make it more likely to pass (I've done this in the past) but if we want 100% success rate, we probably have to make it deterministic. 

If I recall, the issue is the topology. I can't really answer your question @bochaco without knowing the topology of your tests. 

What I mean by this is how are the nodes connected. Do all your nodes connect to every other node? Are they connected linearly, i.e node1 <> Node2 <> Node3 <> Node4, are there clusters etc?

In the multi_hop_propagation tests the topology (IIRC) is pseudo-random. Even if you connect all the peers with each other, the mesh-itself is non-deterministic. i.e the peers form this overlay network of peers via the mesh. When one peer publishes a message, the other peers will only forward to their mesh peers (regardless of how they are connected). This mesh overlay is chosen at random and it can be the case that the mesh doesn't cover all the connected peers. What happens then is the peers outside of the mesh, have to wait for gossip and they can receive the message via gossip. But this requires your tests to wait a few heartbeats to receive gossip. Depending on how connected the mesh is, it can take a few gossip rounds (heartbeats) for the message to propagate to all peers. 

If you just want all the tests to work, you could set the D parameter, `mesh_n_low`, `mesh_n` and `mesh_n_high` such that `mesh_n_low` is your entire peer count. Then the mesh should be the entire test set and messages will always be forwarded. 

The reason we don't do this in the `multi_hop_propagation` test, is because we want it to be `multi-hop`. I.e we don't just forward it to everyone. The topology is such that we have to bounce messages around to test the functionality. 

Hopefully this makes sense. I can go into further details if I have more info :)

**Comment by thomaseizinger** [comment]
> we probably have to make it deterministic.

I guess this would require us to parameterize the RNG source for the entire behaviour?

**Comment by bochaco** [comment]
All that makes sense @AgeManning , I'm trying to clean up some code I have to submit it as a PR with a test which reflects the scenario I described, but I think it's basically as you described it, just pseudo random. 

In addition to that, I've been looking at the implementation of [publish](https://github.com/libp2p/rust-libp2p/blob/master/protocols/gossipsub/src/behaviour.rs#L609) and trying to find the logic corresponding to the documentation I found about the [flood_publish](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md#flood-publishing) feature, which seems to be new and part of v1.1 of the protocol, more specifically about this paragraph:

> In gossipsub v1.1 publishing is (optionally) done by publishing the message to all connected peers with a score above a publish threshold (see [Peer Scoring](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.1.md#peer-scoring) below). This applies regardless of whether the publisher is subscribed to the topic. With flood publishing enabled, the mesh is used when propagating messages from other peers, but a peer's own messages will always be published to all known peers in the topic.


I'll add a change I was playing with to my PR showing where I believe the above described feature/aspect doesn't seem to be in the logic/impl., at least from what I was able to understand this far, I may be wrong for sure. I'll share the PR soon.

**Comment by bochaco** [comment]
I've pushed this [PR](https://github.com/libp2p/rust-libp2p/pull/4587) as a draft with some code I was playing with as mentioned earlier:
1. The first commit is what I think shall fix or cover a missing feature about flood-publishing, i.e. as referenced above: `With flood publishing enabled, the mesh is used when propagating messages from other peers, but a peer's own messages will always be published to all known peers in the topic.`. Does this make sense or what am I missing there?
2. The second commit is a test which reflects the scenario I was asking about before, i.e. one node subscribed and testing it gets all msgs forwarded by all other nodes which are not subscribed to the topic. This test fails quite easily without the 3rd. commit...
3. The 3rd commit is the one I have more doubts about; without it, a node wouldn't forward a msg received/propagated unless it's subscribed itself to the target topic. So I was playing with doing the forwarding in such cases to non-mesh peers if the `flood-publish` is enabled, but I have no idea if that's expected to be part of the v1.1 of the protocol (I don't see that in the doc I linked before), and/or if this would effectively flood the network with msgs and therefore not desirable....?....

I'm not fully sure this is effectively related to the `multi_hop_propagation` test issue, so apologies if I should take this discussion or ask for help on another thread/issue, just let me know.

**Comment by AgeManning** [comment]
Hey @bochaco. 

I'm not sure about the bug you are referring to with flood publishing. We have tests to ensure flood publishing works. See for example here: https://github.com/libp2p/rust-libp2p/blob/master/protocols/gossipsub/src/behaviour/tests.rs#L2105

I'm not really following your first commit. Instead of collecting all the peers on a given topic (the `set` variable), you now collect all peers on all topics. Unless there is not a single peer subscribed to a single topic, this set will never be empty. And then you use that set to send messages to all the peers in the event that we do not have flood_publish enabled. 

I think this commit will then send to all peers when flood publish is not enabled, even those that are not subscribed to the topic. 

Your new test, i imagine would fail often. You setting up 50-100 peers and having only one subscribe to the topic and trying to get messages sent among the nodes. So yeah, this would definately come down to topology. 

The third thing you have added is also not desired I dont think. We shouldn't be sent messages to topics we are not subscribed to. So in any functional environment we shouldn't receive these messages. If one is sent to us by mistake, we shouldn't propagate it onwards, because its an error and potentially a DOS vector. 

I'll comment further on your PR.

**Comment by bochaco** [comment]
Thanks @AgeManning , I've made a couple of comments in reply to yours, but I think I get you there and makes sense to me.

> I'm not really following your first commit. Instead of collecting all the peers on a given topic (the set variable), you now collect all peers on all topics. Unless there is not a single peer subscribed to a single topic, this set will never be empty. And then you use that set to send messages to all the peers in the event that we do not have flood_publish enabled.

Yes, I think my change there is not really correct even for the case I was trying to cover.

But basically the case I was after is that `self.topic_peers.get(&topic_hash)` is `None` when you try to publish a msg from a peer which is not aware of any peer subscribed to that topic (for example in the test I wrote this is what happens). Thus, I thought the **flood-publish** is what would enable such a msg in such case to still be sent to connected peers so it eventually finds its way through. (I'm aware I'm ignoring here what you described about the time required by the heartbeats, which could possibly make my scenario of `topic_peers.get(&topic_hash)` returning `None` less probable)

So perhaps this is just my own confusion, wrong expectation, on the protocol; would it be fair to say that the protocol assumes there has to be enough (some probabilistic minimum number of) subscribers for a topic so publishing a msg on that topic from any peer has real chances of finding its way through?

**Comment by AgeManning** [comment]
Yep, sorry I was going to provide a full explanation yesterday but got distracted. I think there is a confusion about how pubsub systems work here.

The idea is a network of peers get created. The network is formed per-topic. What I mean by this, is if you have 100 peers connected, and all 100 are subscribed to topic "x" and 2 are subscribed to topic "y", then you have a connected network on topic "x" but essentially you only have 2 connected peers on topic "y". Gossipsub (and I think most pubsub systems) partition peers based on their topics. We do not use peers that are subscribed on x to forward any messages of topic y. Doing this would be very computationally intensive and poor design on large systems. There are a number of reasons. 
1. Imagine  a topic "x" that exists that I don't care about, so i don't subscribe to it. If i had to download every message and forward it, it would cost me a lot of bandwidth for no reason. 
2. Any group of peers could create a misc topic and then make every other peer in the network download and forward their messages
3. If i don't care about topic "x", I also shouldn't forward it. Typically messages need to be validated. If someone is sending malicious or broken messages a node should check the message to make sure its valid before sending it on to others, else we would amplify bad/malicious messages. If a node has no idea or concept of topic "x" they cannot validate it because they have no idea what that topic is. 

You should think of peers existing inside their own domain of which topics they subscribe to. 

Your example is intended to fail. If you do not have a connected path of peers subscribed to a topic, your network is disconnected and messages will not propagate between them, this is by design. Your topology (which is per-topic) is disconnected and you're messages wont get routed. You have to make sure you have enough peers subscribed to the topic you want to forward in order for message propagation to work effectively.

**Comment by bochaco** [comment]
Thanks so much @AgeManning for all the help and all the explanation, it's way more clear to me now, it sounds like I was looking to use the protocol in the wrong way.
